plugins/modules/nd_epoch.py

```diff
@@ -0,0 +1,187 @@
+    - When range is selected, max amount epoch IDs to be returned.
+    type: int
+extends_documentation_fragment: cisco.nd.modules
+"""
+
+EXAMPLES = r"""
+- name: Get the latest epoch id
+  cisco.nd.nd_epoch:
+    insights_group: igName
+    site: siteName
+    period: latest
+  register: query_results
+
+- name: Get the epoch id from period
+  cisco.nd.nd_epoch:
+    insights_group: igName
+    site: siteName
+    period: last_week
+  register: period_last_week
+
+- name: Get all epoch ids from last week
+  cisco.nd.nd_epoch:
+    insights_group: igName
+    site: siteName
+    period: last_week
+    range: true
+  register: period_last_week
+
+- name: Get 3 epoch ids from last week closest to latest
+  cisco.nd.nd_epoch:
+    insights_group: igName
+    site: siteName
+    period: last_week
+    range: true
+    max_epochs: 3
+  register: period_last_week
+
+- name: Get all epoch ids from date range
+  cisco.nd.nd_epoch:
+    insights_group: igName
+    site: siteName
+    from_date: 2023-01-01T10:00:00
+    to_date: 2023-01-02T14:00:00
+    range: true
+  register: period_last_week
+"""
+
+RETURN = r"""
+"""
+
+from datetime import datetime as dt, timedelta
+from ansible.module_utils.basic import AnsibleModule
+from ansible_collections.cisco.nd.plugins.module_utils.nd import NDModule, nd_argument_spec
+from ansible_collections.cisco.nd.plugins.module_utils.ndi import NDI
+from ansible_collections.cisco.nd.plugins.module_utils.constants import EPOCH_DELTA_TYPES
+
+
+def main():
+    argument_spec = nd_argument_spec()
+    argument_spec.update(
+        insights_group=dict(type="str", required=True, aliases=["fab_name", "ig_name"]),
+        site=dict(type="str", required=True),
+        period=dict(type="str", choices=list(EPOCH_DELTA_TYPES)),
+        from_date=dict(type="str"),
+        to_date=dict(type="str"),
+        range=dict(type="bool", default=False),
+        max_epochs=dict(type="int"),
+    )
+
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+    )
+
+    nd = NDModule(module)
+    ndi = NDI(nd)
+
+    if not (hasattr(dt, "fromisoformat") and callable(getattr(dt, "fromisoformat"))):
+        nd.fail_json(msg="M(cisco.nd.nd_epoch) can only be used with python 3.7 and higher.")
+
+    insights_group = nd.params.get("insights_group")
+    site = nd.params.get("site")
+    period = nd.params.get("period")
+    from_date = nd.params.get("from_date")
+    to_date = nd.params.get("to_date")
+    max_epochs = nd.params.get("max_epochs")
+
+    if period:
+        to_date = dt.today()
+        from_date = to_date - timedelta(seconds=EPOCH_DELTA_TYPES[period]) if period != "latest" else dt.fromtimestamp(0)
+    else:
+        try:
+            to_date = dt.fromisoformat(to_date) if to_date else dt.today()
+            from_date = dt.fromisoformat(from_date) if from_date else dt.fromtimestamp(0)
+        except ValueError as e:
+            nd.fail_json(msg="{0}".format(e))
```

Comment:
Please add the reason for failure with the failure message?

Reply:
`e` would already be a clear message displayed to the user in my opinion, it will display `Invalid isoformat string` in the msg. This is being captured in a test in the fail test section. What would you like to add to the message?

Reply:
Can you try add some thing like this: "Query epoch data failed due to: error message"

Reply:
Ok will change